### PR TITLE
Add suffix function to the fs module

### PR DESCRIPTION
**Pull request by joukewitteveen**
When iterating over a list of files, it can be convenient to check suffixes. You could already do `fs.name(file).endswith('.txt')`, but `fs.suffix(file) == '.txt'` is just that tiny bit nicer. We also get the invariant `fs.name(file) == fs.stem(file) + fs.suffix(file)`.

### CODE REVIEW

**Comment by joukewitteveen** [comment]
The corner case of an 'empty-but-present suffix' (e.g. `foo/bar.`) will change behavior in [Python 3.14](https://github.com/python/cpython/pull/118952). I've removed our test for it.

File: mesonbuild/modules/fs.py [hunk 1]
```diff
@@ -214,6 +215,13 @@ def stem(self, state: 'ModuleState', args: T.Tuple[T.Union[FileOrString, BuildTa
         path = self._obj_to_path('fs.stem', args[0], state)
         return str(path.stem)
 
+    @noKwargs
+    @typed_pos_args('fs.suffix', (str, File, CustomTarget, CustomTargetIndex, BuildTarget))
+    @FeatureNew('fs.suffix', '1.9.0')
+    def suffix(self, state: 'ModuleState', args: T.Tuple[T.Union[FileOrString, BuildTargetTypes]], kwargs: T.Dict[str, T.Any]) -> str:
```
Comment (dcbaker) [inline mesonbuild/modules/fs.py]:
```suggestion
    def suffix(self, state: ModuleState, args: T.Tuple[T.Union[FileOrString, BuildTargetTypes]], kwargs: T.Dict[str, T.Any]) -> str:
```

Comment (joukewitteveen) [reply mesonbuild/modules/fs.py]: That would be inconsistent with the rest of the file. I can add a commit changing it throughout the file, if you think that is a good idea.

Comment (dcbaker) [reply mesonbuild/modules/fs.py]: Our philosophy on unquoting type declarations has been "new code should do the new thing, if you change old code update it to the current standard." There's a lot of code using quoted type declarations, but updating it creates a lot of useless changes and makes for git rebase conflicts

**Comment by dcbaker** [comment]
Everything looks good to me other than the one very small nit, so this should be good to go.

**Comment by bonzini** [comment]
> The corner case of an 'empty-but-present suffix' (e.g. foo/bar.) will change behavior in Python 3.14.

Should Meson instead use already os.path.splitext so that it already uses the 3.14 behavior?

**Comment by nirbheek** [comment]
> Should Meson instead use already os.path.splitext so that it already uses the 3.14 behavior?

I agree, and also we should have a test for this so the behaviour stays the same regardless of Python version.

**Comment by dcbaker** [comment]
pathlib for concrete paths has a lot of ugly and painful corners (the purepaths are much better), I'd prefer to use `os.path` in general if possible

**Comment by joukewitteveen** [comment]
I'll try to find some time end of next week to redo the module in terms of `os.path` :+1:.

**Comment by joukewitteveen** [comment]
There was a single call to `mesonlib.path_is_in_root` that still required importing `pathlib.Path`. I don't think this PR is the place to change the signature of `path_is_in_root`, so I kept it. Otherwise, the module now directly calls into `os.path`. I used the opportunity to unquote all type specifiers (all function bodies were touched).

**Comment by joukewitteveen** [comment]
The failing checks are due to [a bug in `ntpath.isabs`](https://github.com/python/cpython/issues/44626) that was [fixed](https://github.com/python/cpython/pull/113829) [in Python 3.13](https://docs.python.org/3.13/whatsnew/3.13.html#os-path).

I'm not sure what to do. We cannot use `os.path.splitroot`, because it is only available from Python 3.12 onward. Using a `PurePath` for just this function seems excessive. I'll need to think a bit, suggestions are welcome!

**Comment by bonzini** [comment]
> Using a PurePath for just this function seems excessive

Maybe but I don't see any alternative. Just use a PurePath and add a link to https://github.com/python/cpython/issues/44626 in a comment. Thanks for looking into the failures!

**Comment by joukewitteveen** [comment]
How about this fix? We would otherwise (and currently are) be affected by [a related bug](https://github.com/python/cpython/issues/105476) where incomplete UNC paths (which are absolute because they don't depend on the current working directory) would not be considered absolute.

**Comment by joukewitteveen** [comment]
The previous test failure was due to an implicit assumption that `replace_suffix` changed the alternate path separator on Windows (`/`) to `\`. Regardless of whether we want that, I think the assumption should not be part of the specific test that failed, so I watered that test down (by using `\` both before and after replacement, the test passes with both implementations).

Let me know if we want to reinstate the separator-changing behavior of `replace_suffix`. I think we shouldn't.

**Comment by joukewitteveen** [comment]
The CI failures seem unrelated. Before my last push, where I corrected some authorship details, all checks passed. Maybe someone can restart the unsuccessful checks?

**Comment by jpakkane** [comment]
As your branch is 34 commits behind current master, a better approach is to rebase it. This both reruns the tests and ensures that the MR keeps working against current master (breakage is unlikely, but does happen sometimes).

**Comment by eli-schwartz** [comment]
> pathlib for concrete paths has a lot of ugly and painful corners (the purepaths are much better), I'd prefer to use `os.path` in general if possible

As an aside, @wavyebuilder has informed me of an issue where users were reporting that QEMU could not be built on selinux systems, which turned out to be because pathlib was translating selinux denials of /dev/kvm in unprivileged scopes, into PermissionError tracebacks.

And now I'm only half-jokingly wondering whether to backport this refactor as a standalone bugfix...
